base/mem_trie.hpp

```diff
@@ -1,134 +1,138 @@
 #pragma once
 
 #include "base/macros.hpp"
+#include "base/stl_add.hpp"
 
+#include <cstdint>
 #include <map>
+#include <memory>
 #include <vector>
 
 namespace my
 {
 // This class is a simple in-memory trie which allows to add
 // key-value pairs and then traverse them in a sorted order.
-template <typename TString, typename TValue>
+template <typename String, typename Value>
 class MemTrie
 {
 public:
   MemTrie() = default;
+  MemTrie(MemTrie && rhs) { *this = std::move(rhs); }
 
-  MemTrie(MemTrie && other) : m_root(move(other.m_root))
+  MemTrie & operator=(MemTrie && rhs)
   {
-    m_numNodes = other.m_numNodes;
-    other.m_numNodes = 0;
-  }
-
-  MemTrie & operator=(MemTrie && other)
-  {
-    m_root = move(other.m_root);
-    m_numNodes = other.m_numNodes;
-    other.m_numNodes = 0;
+    m_root = std::move(rhs.m_root);
+    m_numNodes = rhs.m_numNodes;
+    rhs.m_numNodes = 1;
     return *this;
   }
 
   // Adds a key-value pair to the trie.
-  void Add(TString const & key, TValue const & value)
+  void Add(String const & key, Value const & value)
   {
-    Node * cur = &m_root;
+    auto * cur = &m_root;
     for (auto const & c : key)
     {
-      size_t numNewNodes;
-      cur = cur->GetMove(c, numNewNodes);
-      m_numNodes += numNewNodes;
+      bool created;
+      cur = &cur->GetMove(c, created);
+      if (created)
+        ++m_numNodes;
     }
     cur->AddValue(value);
   }
 
   // Traverses all key-value pairs in the trie and calls |toDo| on each of them.
   template <typename ToDo>
-  void ForEach(ToDo && toDo)
+  void ForEachInTrie(ToDo && toDo) const
+  {
+    String prefix;
+    ForEachInSubtree(m_root, prefix, std::forward<ToDo>(toDo));
+  }
+
+  template <typename ToDo>
+  void ForEachInNode(String const & prefix, ToDo && toDo) const
   {
-    TString prefix;
-    ForEach(&m_root, prefix, std::forward<ToDo>(toDo));
+    if (auto const * root = MoveTo(prefix))
+      ForEachInNode(*root, prefix, std::forward<ToDo>(toDo));
   }
 
   template <typename ToDo>
-  void ForEachInSubtree(TString prefix, ToDo && toDo) const
+  void ForEachInSubtree(String prefix, ToDo && toDo) const
```

Comment:
Ah, sorry, no. Prefix is passed by reference to a private recursive implementation of the function, where it's updated. So, we need pass-by-copy here.